src/Microsoft.OData.Core/ODataNotificationWriter.cs

```diff
@@ -324,13 +331,40 @@ protected override void Dispose(bool disposing)
                 if (this.listener != null)
                 {
                     // Tell the listener that the stream is being disposed.
-                    this.listener.StreamDisposed();
+                    if (synchronous)
+                    {
+                        this.listener.StreamDisposed();
+                    }
+                    else
+                    {
+                        this.listener.StreamDisposedAsync().Wait();
+                    }
+
                     this.listener = null;
                 }
             }
 
             this.textWriter.Dispose();
             base.Dispose(disposing);
         }
+
+#if NETSTANDARD2_0
+        public async ValueTask DisposeAsync()
+        {
+            if (this.listener != null)
+            {
+                await this.listener.StreamDisposedAsync()
+                    .ConfigureAwait(false);

```

Comment:
Is it possible that the listener only implemented the synchronous version `StreamDisposed`?

Reply:
@habbes Previously yes, we only had the asynchronous `StreamDisposed` method. In `ODataJsonLightWriter` for example, here is how that method was implemented:
```csharp
        void IODataStreamListener.StreamDisposed()
        {
            // Complete writing the stream
            if (this.State == WriterState.Stream)
            {
                this.EndBinaryStream();
            }
            else if (this.State == WriterState.String)
            {
                this.EndTextWriter();
            }
            this.LeaveScope();
        }
```
You'll find that the method called from here actually perform some writing. Given that we want to achieve true async, that required introduction of an asynchronous equivalent - `StreamDisposedAsync` - looking something like this:
```csharp
        async Task IODataStreamListener.StreamDisposedAsync()
        {
            // Complete writing the stream
            if (this.State == WriterState.Stream)
            {
                await this.EndBinaryStreamAsync()
                    .ConfigureAwait(false);
            }
            else if (this.State == WriterState.String)
            {
                await this.EndTextWriterAsync()
                    .ConfigureAwait(false);
            }

            await this.LeaveScopeAsync()
                .ConfigureAwait(false);
        }
```
This therefore is intended to help us deliver the promise of a true asynchronous API.

Reply:
Oh, so they will be required to implement `StreamDisposedAsync()` otherwise get a compiler error?

Reply:
An earlier PR introduced the new method with a body that just throws a `NotImplementedException`. In subsequent PRs, I have been replacing that with the actual implementation